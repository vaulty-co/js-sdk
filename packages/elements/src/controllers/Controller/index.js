import invariant from 'invariant';
import { BroadcastChannel } from 'broadcast-channel';
import { SlaveChannel } from '@js-sdk/common/src/channels/SlaveChannel';
import { isSafari } from '@js-sdk/common/src/helpers/isSafari';

import { Config } from '../../config';

const CONTROLLER_STATUSES = {
  INIT: 'init',
  DESTROYED: 'destroyed',
};

/**
 * Controller base class
 * @class
 * @extends {ElementsInstance}
 */
class Controller {
  static get STATUSES() {
    return CONTROLLER_STATUSES;
  }

  /**
   * @param {boolean} condition
   * @param {string} message
   */
  static invariant(condition, message) {
    invariant(condition, `[${this.name}] ${message}`);
  }

  /**
   * @param {string} options.channelId
   * @param {string} options.sdkId
   */
  constructor(options) {
    this.status = CONTROLLER_STATUSES.INIT;
    this.channelId = options?.channelId;

    let broadcastChannelOptions = {};
    // Notice: Safari does not allow IndexDB fallback in iFrame and we should manually
    // use localStorage
    if (isSafari()) {
      broadcastChannelOptions = {
        type: 'localstorage',
      };
    }
    this.broadcastChannel = new BroadcastChannel(
      options?.sdkId,
      broadcastChannelOptions,
    );
  }

  /**
   * Mount controller
   */
  mount() {
    this.openChannel();
    this.registerHandlers();
  }

  /**
   * Unmount controller
   */
  unmount() {
    this.destroy();
  }

  /**
   * Open channel with master frame
   * @protected
   */
  openChannel() {
    this.constructor.invariant(
      Boolean(this.channelId),
      'Controller does not have specified "channelId" for opening channel',
    );

    this.controllerSlaveChannel = new SlaveChannel({
      channelId: this.channelId,
      targetOrigin: Config.sdkOrigin,
    });
    this.controllerSlaveChannel.connect();
  }

  /**
   * Register allowed messages by master frame and give response by this messages
   * @private
   */
  // eslint-disable-line class-methods-use-this
  registerHandlers() {

  }

  /**
   * Destroy controller
   */
  destroy() {
    if (this.status === CONTROLLER_STATUSES.DESTROYED) {
      this.constructor.invariant(
        false,
        'Controller is destroyed and can not be destroyed again.',
      );
      return;
    }

    this.controllerSlaveChannel.destroy();
    this.controllerSlaveChannel = null;

    this.broadcastChannel.close();
    this.broadcastChannel = null;
  }
}

export default Controller;
export {
  Controller,
};
