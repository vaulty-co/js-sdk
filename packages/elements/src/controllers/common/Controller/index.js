import { SlaveChannel } from '@vaulty/common/src/channels/SlaveChannel';
import { createBroadcastChannel } from '@vaulty/common/src/channels/utils';
import { staticInvariant } from '@vaulty/common/src/helpers/invariant';
import { Message } from '@vaulty/common/src/channels/Message';

import { Config } from '../../../config';
import {
  PUT_CONTROLLER_REQUEST,
  ADD_FIELDS_IDS_TO_CONTROLLER_REQUEST,
  REMOVE_FIELDS_IDS_FROM_CONTROLLER_REQUEST,
  PATCH_CONTROLLER_STATUS_REQUEST,
  CONTROLLER_LOADED_WATCHER,
} from './messages';
import { putControllerHandler } from './handlers/putController';
import { addFieldsIdsHandler } from './handlers/addFields';
import { removeFieldsIdsHandler } from './handlers/removeFields';
import { patchStatusHandler } from './handlers/patchStatus';

const CONTROLLER_STATUSES = {
  INIT: 'init',
  DESTROYED: 'destroyed',
};

/**
 * Controller base class
 * @class ControllerElement
 * @extends ElementsInstance
 */
class Controller {
  static get STATUSES() {
    return CONTROLLER_STATUSES;
  }

  static get invariant() {
    return staticInvariant;
  }

  /**
   * @param {string} options.channelId
   * @param {string} options.sdkId
   */
  constructor(options) {
    this.status = CONTROLLER_STATUSES.INIT;
    this.channelId = options?.channelId;
    /**
     * @type {?ControllerModel}
     */
    this.controllerModel = null;
    this.broadcastChannel = createBroadcastChannel(options?.sdkId);
  }

  /**
   * Mount controller
   */
  mount() {
    this.openChannel();
    this.registerHandlers();
  }

  /**
   * Unmount controller
   */
  unmount() {
    this.destroy();
  }

  /**
   * Open channel with master frame
   * @protected
   */
  openChannel() {
    this.constructor.invariant(
      Boolean(this.channelId),
      'Controller does not have specified "channelId" for opening channel',
    );

    this.controllerSlaveChannel = new SlaveChannel({
      channelId: this.channelId,
      targetOrigin: Config.sdkOrigin,
    });
    this.controllerSlaveChannel.connect();
    this.controllerSlaveChannel.postMessage(
      new Message(CONTROLLER_LOADED_WATCHER),
    );
  }

  /**
   * Register allowed messages by master frame and give response by this messages
   * @protected
   */
  registerHandlers() {
    this.controllerSlaveChannel.subscribe(
      PUT_CONTROLLER_REQUEST,
      putControllerHandler.bind(this),
    );
    this.controllerSlaveChannel.subscribe(
      ADD_FIELDS_IDS_TO_CONTROLLER_REQUEST,
      addFieldsIdsHandler.bind(this),
    );
    this.controllerSlaveChannel.subscribe(
      REMOVE_FIELDS_IDS_FROM_CONTROLLER_REQUEST,
      removeFieldsIdsHandler.bind(this),
    );
    this.controllerSlaveChannel.subscribe(
      PATCH_CONTROLLER_STATUS_REQUEST,
      patchStatusHandler.bind(this),
    );
  }

  /**
   * Destroy controller
   */
  destroy() {
    if (this.status === CONTROLLER_STATUSES.DESTROYED) {
      this.constructor.invariant(
        false,
        'Controller is destroyed and can not be destroyed again.',
      );
      return;
    }

    this.controllerSlaveChannel.destroy();
    this.controllerSlaveChannel = null;

    this.broadcastChannel.close();
    this.broadcastChannel = null;

    this.controllerModel = null;
  }
}

export default Controller;
export {
  Controller,
};
