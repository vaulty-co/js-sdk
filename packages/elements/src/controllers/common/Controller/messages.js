/*
 * Requests
 */
const PUT_CONTROLLER_REQUEST = 'put_controller_request';
const PUT_CONTROLLER_RESPONSE = 'put_controller_response';
const ADD_FIELDS_IDS_TO_CONTROLLER_REQUEST = 'add_fields_ids_to_controller_request';
const ADD_FIELDS_IDS_TO_CONTROLLER_RESPONSE = 'add_fields_ids_to_controller_response';
const REMOVE_FIELDS_IDS_FROM_CONTROLLER_REQUEST = 'remove_fields_ids_from_controller_request';
const REMOVE_FIELDS_IDS_FROM_CONTROLLER_RESPONSE = 'remove_fields_ids_from_controller_response';
const PATCH_CONTROLLER_STATUS_REQUEST = 'patch_controller_status_request';
const PATCH_CONTROLLER_STATUS_RESPONSE = 'patch_controller_status_response';

/*
 * Watchers
 */
const CONTROLLER_LOADED_WATCHER = 'controller_loaded_watcher';


export {
  PUT_CONTROLLER_REQUEST,
  PUT_CONTROLLER_RESPONSE,
  PATCH_CONTROLLER_STATUS_REQUEST,
  PATCH_CONTROLLER_STATUS_RESPONSE,
  ADD_FIELDS_IDS_TO_CONTROLLER_REQUEST,
  ADD_FIELDS_IDS_TO_CONTROLLER_RESPONSE,
  REMOVE_FIELDS_IDS_FROM_CONTROLLER_REQUEST,
  REMOVE_FIELDS_IDS_FROM_CONTROLLER_RESPONSE,
  CONTROLLER_LOADED_WATCHER,
};
