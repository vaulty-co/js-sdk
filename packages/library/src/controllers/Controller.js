// FIXME - disabling eslint should be removed after creating behavior for methods
/* eslint-disable class-methods-use-this */
import invariant from 'invariant';
import EventEmitter from 'events';

import { NODE_TYPES } from '../constants/nodeTypes';

/**
 * @class
 */
class Controller {
  /**
   * @param {boolean} condition
   * @param {string} message
   */
  static invariant(condition, message) {
    invariant(condition, `[${this.name}] ${message}`);
  }

  constructor() {
    /**
     * @type {EventEmitter}
     */
    this.events = new EventEmitter();
  }

  /**
   * Append controller in some DOM node
   * @param {string|HTMLElement} node - valid css selector or DOM node, where element should be appended
   */
  appendTo(node) {
    // TODO - it should be some util
    let resultNode = node;
    if (typeof node === 'string') {
      resultNode = document.querySelector(node);
    }
    this.parent = node;
    this.constructor.invariant(
      this.controllerIframe,
      'Controller should contains "controllerIframe" property for manipulating its',
    );
    this.constructor.invariant(
      typeof resultNode === 'object'
      // TODO - NODE_TYPE should be placed in utils
      && resultNode.nodeType === NODE_TYPES.ELEMENT_NODE,
      'Controller should be "appendTo" HTMLElement',
    );
    this.controllerIframe.appendTo(this.parent);
  }

  /**
   * Add handler to event
   * @param {string} eventName
   * @param {Function} eventHandler
   * @return {*}
   */
  on(eventName, eventHandler) {
    return this.events.on(eventName, eventHandler);
  }

  /**
   * Remove handler from event
   * @param {string} eventName
   * @param {Function} eventHandler
   * @return {*}
   */
  off(eventName, eventHandler) {
    return this.events.off(eventName, eventHandler);
  }

  /**
   * Destroy controller and its DOM tree. It does not destroy parent, where controller have been placed.
   * Controller is not usable after destroy
   */
  destroy() {
    this.controllerIframe.destroy();

    this.parent = null;
    this.controllerIframe = null;

    this.events.removeAllListeners();
  }
}

export default Controller;
export {
  Controller,
};
