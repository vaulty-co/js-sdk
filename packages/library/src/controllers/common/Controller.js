import EventEmitter from 'events';
import { staticInvariant } from '@vaulty/common/src/helpers/invariant';
import uniqueId from '@vaulty/common/src/helpers/uniqueId';

import { actionsToDispatch } from '../../store/utils/actionsToDispatch';
import { makeControllerSelector, makeControllerStatusSelector } from './store/selectors';
import { operationCreateController } from './store/operations/createController';
import { operationMountController } from './store/operations/mountController';
import { operationAddFieldsToController } from './store/operations/addFieldsToController';
import { operationRemoveFieldsFromController } from './store/operations/removeFieldsFromController';
import { operationDestroyController } from './store/operations/destroyController';

/**
 * Get fields ids by field instances
 * @param {Array<Field>} [fields = []]
 * @return {Array<string>}
 */
const getFieldsIds = (fields = []) => (
  fields.map(
    (field) => field.id,
  )
);

/**
 * @class SDKController
 */
class Controller {
  static get invariant() {
    return staticInvariant;
  }

  /**
   * @param {SDKControllerOptions|SDKFormOptions} options
   */
  constructor(options) {
    /**
     * @type {string}
     * @protected
     */
    this.id = uniqueId();
    /**
     * @type {SDKControllerOptions}
     * @protected
     */
    this.options = options;
    /**
     * @type {?EventEmitter}
     * @protected
     */
    this.events = new EventEmitter();
    /**
     * @type {?IFrameNode}
     * @protected
     */
    this.controllerIframe = null;
    /**
     * Controller selector
     * @protected
     */
    this.controllerSelector = makeControllerSelector(this.id);
    /**
     * Controller status selector
     * @protected
     */
    this.controllerStatusSelector = makeControllerStatusSelector(this.id);

    /**
     * @type {Object.<string, operation>}
     * @protected
     */
    this.operations = actionsToDispatch(options.store)({
      createController: operationCreateController,
      mountController: operationMountController,
      addFields: operationAddFieldsToController,
      removeFields: operationRemoveFieldsFromController,
      destroyController: operationDestroyController,
    });
    this.operations.createController({
      id: this.id,
      fieldsIds: getFieldsIds(options.fields),
    });

    const state = options.store.getState();
    let currentStatus = this.controllerStatusSelector(state);
    const unsubscribe = options.store.subscribe(() => {
      const newState = options.store.getState();
      const currentController = this.controllerSelector(newState);
      if (!currentController) {
        this.controllerIframe.destroy();
        this.controllerIframe = null;
        this.events.removeAllListeners();
        this.events = null;
        unsubscribe();
      } else {
        const previousStatus = currentStatus;
        currentStatus = this.controllerStatusSelector(newState);
        if (previousStatus !== currentStatus) {
          this.events.emit('status', currentStatus);
        }
      }
    });
  }

  /**
   * Add fields to controller
   * @param {Array<Field>} fields
   */
  addFields(fields) {
    this.operations.addFields(
      getFieldsIds(fields),
    );
  }

  /**
   * Remove fields from controller
   * @param {Array<Field>} fields
   */
  removeFields(fields) {
    this.operations.removeFields(
      getFieldsIds(fields),
    );
  }

  /**
   * Remove fields from controller
   */
  removeAllFields() {
    const controller = this.controllerSelector(this.options.store.getState());
    this.operations.removeFields(controller.fieldsIds);
  }

  /**
   * Append controller in some DOM node
   * @param {string|HTMLElement} parentNode - valid css selector or DOM node, where element should be appended
   */
  appendTo(parentNode) {
    this.constructor.invariant(
      Boolean(this.controllerIframe),
      'IFrameNode should exist in controller',
    );

    this.controllerIframe.appendTo(parentNode);
    this.operations.mountController({
      id: this.id,
      iframeDomNode: this.controllerIframe.node,
    });
  }

  /**
   * Add handler to event
   * @param {string} eventName
   * @param {Function} eventHandler
   * @return {*}
   */
  on(eventName, eventHandler) {
    return this.events.on(eventName, eventHandler);
  }

  /**
   * Remove handler from event
   * @param {string} eventName
   * @param {Function} eventHandler
   * @return {*}
   */
  off(eventName, eventHandler) {
    return this.events.off(eventName, eventHandler);
  }

  /**
   * Get controller status
   * Notice: real readiness is computed value by form readiness and fields readiness
   * @returns {ComputedControllerStatus}
   */
  getStatus() {
    return this.controllerStatusSelector(this.options.store.getState());
  }

  /**
   * Destroy controller and its DOM tree. It does not destroy parent, where controller have been placed.
   * Controller is not usable after destroy
   * @param {boolean} [withFields = false] - destroy controller with its fields
   */
  destroy(withFields) {
    if (typeof withFields === 'undefined') {
      /* eslint no-param-reassign:0 */
      withFields = false;
    }
    this.operations.destroyController({ id: this.id, withFields });
  }
}

export default Controller;
export {
  Controller,
};
