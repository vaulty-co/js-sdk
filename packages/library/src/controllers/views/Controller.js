import EventEmitter from 'events';
import { staticInvariant } from '@js-sdk/common/src/helpers/invariant';

/**
 * @class
 */
class Controller {
  static get invariant() {
    return staticInvariant;
  }

  constructor() {
    /**
     * @type {EventEmitter}
     */
    this.events = new EventEmitter();
  }

  /**
   * Append controller in some DOM node
   * @param {string|HTMLElement} parentNode - valid css selector or DOM node, where element should be appended
   */
  appendTo(parentNode) {
    this.controllerIframe.appendTo(parentNode);
  }

  /**
   * Add handler to event
   * @param {string} eventName
   * @param {Function} eventHandler
   * @return {*}
   */
  on(eventName, eventHandler) {
    return this.events.on(eventName, eventHandler);
  }

  /**
   * Remove handler from event
   * @param {string} eventName
   * @param {Function} eventHandler
   * @return {*}
   */
  off(eventName, eventHandler) {
    return this.events.off(eventName, eventHandler);
  }

  /**
   * Destroy controller and its DOM tree. It does not destroy parent, where controller have been placed.
   * Controller is not usable after destroy
   */
  destroy() {
    this.controllerIframe.destroy();
    this.controllerIframe = null;

    this.events.removeAllListeners();
  }
}

export default Controller;
export {
  Controller,
};
