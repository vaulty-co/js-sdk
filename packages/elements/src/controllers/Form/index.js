// FIXME - disabling eslint should be removed after creating behavior for methods
/* eslint-disable class-methods-use-this */
import { Controller } from '../Controller';

/**
 * Form controller for submitting data from fields
 */
class Form extends Controller {
  /**
   * Submit data from fields
   */
  submit() {
  }
}

export default Form;
export {
  Form,
};
