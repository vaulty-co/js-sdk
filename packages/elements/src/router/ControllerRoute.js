import { Route } from './Route';
import { Controller } from '../controllers/Controller';

/**
 * Route for Controller
 */
class ControllerRoute extends Route {
  /**
   * Render route
   * @returns {Controller}
   */
  render() {
    const controller = this.renderFunction();
    this.constructor.invariant(
      (controller instanceof Controller),
      `Controller route "${this.name}" does not return Controller in render function.`,
    );
    return controller;
  }
}

export default ControllerRoute;
export {
  ControllerRoute,
};
