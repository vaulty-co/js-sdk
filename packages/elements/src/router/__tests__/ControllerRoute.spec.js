import { Controller } from '@js-sdk/elements/src/controllers/Controller';
import { ControllerRoute } from '../ControllerRoute';

describe('ControllerRoute', () => {
  let route;
  let renderFunction;

  beforeEach(() => {
    renderFunction = jest.fn(() => new Controller());
    route = new ControllerRoute(
      'from',
      {
        controller: 'form',
      },
      renderFunction,
    );
  });

  describe('#render', () => {
    it('should run render function', () => {
      route.render();

      expect(renderFunction).toBeCalled();
    });

    it('should throw exception, when it does not return Controller instance', () => {
      route = new ControllerRoute(
        'form',
        {
          controller: 'form',
        },
        () => null,
      );

      expect(() => {
        route.render();
      }).toThrowErrorMatchingSnapshot();
    });
  });
});
