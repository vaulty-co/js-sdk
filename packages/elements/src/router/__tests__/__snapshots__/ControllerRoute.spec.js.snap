// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControllerRoute #render should throw exception, when it does not return Controller instance 1`] = `"[ControllerRoute] Controller route \\"form\\" does not return Controller in render function."`;
